internal/asm/f32/ge_amd64.s

```diff
@@ -0,0 +1,773 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//+build go1.8,!noasm,!appengine
+
+#include "textflag.h"
+
+#define SIZE 4
+#define BITSIZE 2
+#define KERNELSIZE 3
+
+#define M_DIM m+0(FP)
+#define M CX
+#define N_DIM n+8(FP)
+#define N BX
+
+#define TMP1 R14
+#define TMP2 R15
+
+#define X_PTR SI
+#define Y y_base+56(FP)
+#define Y_PTR DX
+#define A_ROW AX
+#define A_PTR DI
+
+#define INC_X R8
+#define INC3_X R9
+
+#define INC_Y R10
+#define INC3_Y R11
+
+#define LDA R12
+#define LDA3 R13
+
+#define ALPHA X0
+#define ALPHA_SPILL al-16(SP)
+
+#define LOAD_ALPHA \
+	MOVSS  alpha+16(FP), ALPHA \
+	SHUFPS $0, ALPHA, ALPHA
+
+#define LOAD4 \
+	PREFETCHNTA 16*SIZE(X_PTR )   \
+	MOVDDUP     (X_PTR), X1       \
+	MOVDDUP     2*SIZE(X_PTR), X3 \
+	MOVSHDUP    X1, X2            \
+	MOVSHDUP    X3, X4            \
+	MOVSLDUP    X1, X1            \
+	MOVSLDUP    X3, X3            \
+	MULPS       ALPHA, X1         \
+	MULPS       ALPHA, X2         \
+	MULPS       ALPHA, X3         \
+	MULPS       ALPHA, X4
+
+#define LOAD2 \
+	MOVDDUP  (X_PTR), X1 \
+	MOVSHDUP X1, X2      \
+	MOVSLDUP X1, X1      \
+	MULPS    ALPHA, X1   \
+	MULPS    ALPHA, X2
+
+#define LOAD1 \
+	MOVSS  (X_PTR), X1 \
+	SHUFPS $0, X1, X1  \
+	MULPS  ALPHA, X1
+
+#define LOAD4_INC \
+	PREFETCHNTA (X_PTR )(INC_X*8)     \
+	MOVSS       (X_PTR), X1           \
+	MOVSS       (X_PTR)(INC_X*1), X2  \
+	MOVSS       (X_PTR)(INC_X*2), X3  \
+	MOVSS       (X_PTR)(INC3_X*1), X4 \
+	SHUFPS      $0, X1, X1            \
+	SHUFPS      $0, X2, X2            \
+	SHUFPS      $0, X3, X3            \
+	SHUFPS      $0, X4, X4            \
+	MULPS       ALPHA, X1             \
+	MULPS       ALPHA, X2             \
+	MULPS       ALPHA, X3             \
+	MULPS       ALPHA, X4
+
+#define LOAD2_INC \
+	MOVSS  (X_PTR), X1          \
+	MOVSS  (X_PTR)(INC_X*1), X2 \
+	SHUFPS $0, X1, X1           \
+	SHUFPS $0, X2, X2           \
+	MULPS  ALPHA, X1            \
+	MULPS  ALPHA, X2
+
+#define KERNEL_LOAD8 \
+	MOVUPS (Y_PTR), X5       \
+	MOVUPS 4*SIZE(Y_PTR), X6
+
+#define KERNEL_LOAD8_INC \
```

Comment:
This is pretty scary to read (I miss my old 6502 and Z80). It might be nice to have a broad overview comment explaining these convoluting KERNEL* macros.